texar/modules/pretrained/xlnet_base.py

```diff
@@ -0,0 +1,96 @@
+# Copyright 2019 The Texar Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Base class for XLNet Modules.
+"""
+
+from typing import Optional
+
+from texar.hyperparams import HParams
+from texar.module_base import ModuleBase
+from texar.modules.pretrained import xlnet_utils
+
+
+__all__ = [
+    "XLNetBase",
+]
+
+
+class XLNetBase(ModuleBase):
```

Comment:
Just a thought - Why do we need XLNetBase and BertBase classes? Can we not have XLNetEncoder and BERTEncoder inherit from EncoderBase and avoid this inheritance? It seems to this, this hierarchy may be a bit unnecessary.

Reply:
I think the point of having `XLNetBase` (and `BertBase` similarly) is to separate the logic for loading pretrained models. That said, I feel the loading logic is the same for almost all pretrained models, so maybe we can unify them as something like `PretrainedBase`?

Reply:
Because I here want to separate model downloading logic from the actual model building logic. What's more, there might be some modified version of the encoder in the future, and they can share the same base class. @AvinashBukkittu

Reply:
Hmm. What seems weird to me is that neither `XLNetBase` nor `XLNetEncoder` inherits from `EncoderBase` (although I know `EncoderBase` is an empty class essentially.) This is unlike other encoders we have in `texar`. At this point, I am thinking to have `XLNetEncoder` inherit from `XLNetBase` and `EncoderBase` (in this way we can still separate the logic of downloading the model and building the model.)

Reply:
What is the benefit to inherit from `EncoderBase`? @AvinashBukkittu

Reply:
I think the benefit would be a clear inheritance hierarchy, and we can check whether something is a decoder by `isinstance(x, EncoderBase)`.
This might require us to redesign constructor interfaces for `DecoderBase`, `EncoderBase`, and maybe `ClassifierBase`, so they allow cooperative inheritance.

Reply:
If in the future, we want to add some additional properties to all the encoders, we can do it at one place i.e `EncoderBase`. Currently it is not possible for `XLNetEncoder` and `BertEncoder` as they don't inherit from the base class.

Reply:
Yes, I think we should have a `PretrainedBase` instead of the different base classes for different pre-trained models.

Reply:
I agree that all encoders should inherit `EncoderBase`

For XLNet, now `XLNetDecoder` inherits `XLNetEncoder`. Is it a good structure? Or perhaps a better way is to put shared things to `XLNetBase`, and both `XLNetDecoder` and `XLNetEncoder` inherit it?

Reply:
Structure-wise, `XLNetDecoder` and `XLNetEncoder` are pretty similar. `XLNetDecoder` just has one more bias parameter. Here are my thoughts on the `pretrained` modules:

We have a general `PretrainedBase` and multiple utils files including `pretrained_utils.py`, `bert_utils.py`, `gpt2_utils.py`, and `xlnet_utils.py` (maybe also `xlnet_model_utils.py`). `pretrained_utils.py` contains some shared functions used in `xxx_utils.py`. `PretrainedBase` will call the corresponding `util` function according to the class name (Bertxxx, GPT2xxx, XLNetxxx), and every pretrained_encoder class will inherits `PretrainedBase`.

Reply:
I think calling based on the class name is not a good design pattern. We can add a class attribute for each of the pretrained models, e.g. `BertEncoder.model_name = "BERT`, and `PretrainedBase` can lookup required functions based on that.

BTW, I think `BertXXX` should be renamed to `BERTXXX`.